geps/gep-1731/index.md

```diff
@@ -0,0 +1,427 @@
+// 600-999 (inclusive) range, which are not valid for HTTP clients, but are
+// sometimes used for communicating application-specific errors.
+//
+// Implementations MAY support additional shorthand codes for any `[0-9]xx`
+// range.
+//
+// +kubebuilder:validation:Pattern=`^[1-9](?:[0-9][0-9]|xx)$`
+type HTTPRouteRetryStatusCode string
+
+// Duration is a string value representing a duration in time. The foramat is
+// as specified in GEP-2257, a strict subset of the syntax parsed by Golang
+// time.ParseDuration.
+//
+// +kubebuilder:validation:Pattern=`^([0-9]{1,5}(h|m|s|ms)){1,4}$`
+type Duration string
+```
+
+### YAML
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1
+kind: HTTPRoute
+metadata:
+  name: retry-example
+spec:
+  ...
+  rules:
+  - backendRefs:
+    - name: some-service
+      port: 8080
+      retry:
+        codes:
+        - 5xx
+        attempts: 2
+        backoff: 100ms
+```
+
+## Conformance Details
+
+Basic support for configuring retries in HTTPRoute up to a specified maximum count and with a minimum duration between attempts will be gated on the `SupportHTTPRouteRetry` feature. Retries will be tested in combination with request and backend timeouts if supported by the implementation.
+
+Retrying requests based on HTTP status codes will be gated under the following features:
+
+* `SupportHTTPRRouteRetryCodes`
+  * Only 500, 502, 503 and 504 will be tested for conformance.
+  * Traefik does not seem to support specifying error codes, and will only retry on backend timeouts.
+* `SupportHTTPRRouteRetryCode5xx`
+  * 500, 502, 503 and 504 will all be tested for conformance.
+  * Arbitrary status codes in the 500-599 (inclusive) range will not be tested for conformance.
+
+Implementations MAY support specifying additional individual error codes in the valid 100-599 (inclusive) range, invalid 600-999 (inclusive) range or prefix shorthands for any `[0-9]xx` range, but none of these will be tested in conformance.
+
+## Alternatives
+
+### Standard RetryPolicy metaresource
+
+This may be a reasonable approach for configuring broad default retry policies, but the UX of using a separate resource could feel overly burdensome for basic use cases or granular configuration, and the structure of the policy attachment model is [currently being revised](https://github.com/kubernetes-sigs/gateway-api/discussions/2927). Additionally, multiple data plane proxies warn in their documentation about the danger of configuring broad retry policies which may cause logical application errors from replaying non-idempotent requests or overwhelm backends. This approach may still be considered to enable future goals.
+
+### HTTPRoute filter
+
+TODO
+
+### What accommodations are needed for future retry budget support?
+
+Should the `retry` stanza follow the Kubernetes "tagged union" pattern with something like a `mode: "count"` to allow future design space for `mode: "budget"` with distinct sibling fields?
```

Comment:
Seeking feedback on this.

Reply:
We _can_ add this later without it being a breaking change as long as `mode: count` is defaulted if not present, that's not a breaking change. I think it's likely we'll use it for budgeted retries in the future, but that means we can probably get away with not adding it now.

Reply:
This is especially true since it seems like `attempts` will be mutually exclusive with `budget` or whatever.

Reply:
Agree with the outcome of this discussion, just would recommend capturing this in the GEP.